tests/AppTest.php

```diff
@@ -68,4 +70,141 @@ public function testUnexpectedOutputLateError(): void
+            [['test/index', 'page_args' => 'atk4'], false, ['extra_args' => 'atk4'], '/test/?extra_args=atk4&query_args=atk4&page_args=atk4', 'test/index.php?extra_args=atk4&query_args=atk4&page_args=atk4', 'test/index.php8?extra_args=atk4&query_args=atk4&page_args=atk4', 'test/index?extra_args=atk4&query_args=atk4&page_args=atk4'],
+            [['test', 'page_args' => 'atk4'], true, ['extra_args' => 'atk4'], '/', '/index.php?extra_args=atk4', '/default.php8?extra_args=atk4', '/?extra_args=atk4'],
+            [['test', 'page_args' => 'atk4'], true, ['extra_args' => 'atk4'], '/request-url', '/request-url.php?extra_args=atk4', '/request-url.php8?extra_args=atk4', '/request-url?extra_args=atk4'],
+            [['test', 'page_args' => 'atk4'], true, ['extra_args' => 'atk4'], '/request-url/', '/request-url/index.php?extra_args=atk4', '/request-url/default.php8?extra_args=atk4', '/request-url/?extra_args=atk4'],
+            [['test/index', 'page_args' => 'atk4'], true, ['extra_args' => 'atk4'], '/test/', '/test/index.php?extra_args=atk4', '/test/default.php8?extra_args=atk4', '/test/?extra_args=atk4'],
+            [['test', 'page_args' => 'atk4', 'check_unset_page' => false], false, ['extra_args' => 'atk4'], '/?extra_args=atk4&query_args=atk4&page_args=atk4', 'test.php?extra_args=atk4&query_args=atk4&page_args=atk4', 'test.php8?extra_args=atk4&query_args=atk4&page_args=atk4', 'test?extra_args=atk4&query_args=atk4&page_args=atk4'],
+        ];
+    }
+
+    /**
+     * @dataProvider provideUrlBuildingCases
+     *
+     * @param string|array<0|string, string|int|false> $page                URL as string or array with page name as first element and other GET arguments
+     * @param bool                                     $useRequestUrl       Simply return $_SERVER['REQUEST_URI'] if needed
+     * @param array<string, string>                    $extraRequestUrlArgs Additional URL arguments, deleting sticky can delete them
+     */
+    public function testUrlBuilding($page, bool $useRequestUrl, array $extraRequestUrlArgs, string $requestUrl, string $exceptedStd, string $exceptedCustom, string $exceptedRouting): void
+    {
+        $factory = new Psr17Factory();
+        $request = $factory->createRequest('GET', 'http://127.0.0.1' . $requestUrl);
+
+        $_SERVER = [
+            'REQUEST_METHOD' => 'GET',
+            'HTTP_HOST' => $request->getUri()->getHost(),
+            'REQUEST_URI' => $requestUrl,
+            'QUERY_STRING' => $request->getUri()->getQuery(),
+        ];
+
+        parse_str($request->getUri()->getQuery(), $_GET);
+
+        foreach ($extraRequestUrlArgs as $key => $value) {
+            $_GET[$key] = $value;
+        }
+
+        $stickyGetArguments = [
+            '__atk_json' => false,
+            '__atk_tab' => false,
+        ];
+
+        foreach ($_GET as $key => $value) {
+            $stickyGetArguments[$key] = $value;
+        }
+
+        $app = new App([
```

Comment:
AppTest has `CreateAppTrait` trait, use it :)

Reply:
i think is better than createApp + setDefaults after __construct call, no?

Reply:
Ok done